queenbee/io/function.py

```diff
@@ -0,0 +1,222 @@
+
+    Note that values that evaluate to true or false, such as 1 and 0, are not accepted.
+
+    You can add additional validation by defining a JSONSchema specification.
+
+    See http://json-schema.org/understanding-json-schema/reference/boolean.html for more
+    information.
+    """
+    type: constr(regex='^FunctionBooleanInput$') = 'FunctionBooleanInput'
+
+
+class FunctionFolderInput(DAGFolderInput):
+    """A folder input.
+
+    Folder is a special string input. Unlike other string inputs, a folder will be copied
+    from its location to execution folder when a workflow is executed.
+
+    You can add additional validation by defining a JSONSchema specification.
+
+    See http://json-schema.org/understanding-json-schema/reference/string.html#string for
+    more information.
+
+    .. code-block:: python
+
+        "schema": {
+            "type": "string",
+            "maxLength": 50,
+        }
+    """
+    type: constr(regex='^FunctionFolderInput$') = 'FunctionFolderInput'
+
+    # TODO: Change this path to target_path and the one for output to source_path
+    # for this iteration I'm keeping the changes as minimum as possible to break as
+    # little as possible.
+    path: str = Field(
+        ...,
+        description='Path to the target location that the input will be copied to. '
+        ' This path is relative to the working directory where the command is executed.'
+    )
+
+
+class FunctionFileInput(FunctionFolderInput):
+    """A file input.
+
+    File is a special string input. Unlike other string inputs, a file will be copied
+    from its location to execution folder when a workflow is executed.
+
+    You can add additional validation by defining a JSONSchema specification.
+
+    See http://json-schema.org/understanding-json-schema/reference/string.html#string for
+    more information.
+
+    .. code-block:: python
+
+        # a file with maximum 50 characters with an ``epw`` extension.
+
+        "schema": {
+            "type": "string",
+            "maxLength": 50,
+            "pattern": "(?i)(^.*\\.epw$)"
+        }
+
+    """
+    type: constr(regex='^FunctionFileInput$') = 'FunctionFileInput'
+
+    extensions: List[str] = Field(
+        None,
+        description='Optional list of extensions for file. The check for extension is '
+        'case-insensitive.'
+    )
+
+    def validate_spec(self, value):
+        """Validate an input value against specification.
+
+        Use this for validating workflow inputs against a recipe.
+        """
+        assert os.path.isfile(value), f'There is no file at {value}'
+        if self.extensions:
+            assert value.lower().endswith(self.extension.lower()), \
+                f'Input file extension for {value} must be {self.extensions}'
+        if self.spec:
+            spec = dict(self.spec)
+            spec['type'] = 'string'
+            json_schema_validator(value, spec)
+
+
+class FunctionPathInput(FunctionFileInput):
```

Comment:
The current hierarchy is **FunctionFolderInput** => **FunctionFileInput** => **FunctionPathInput**.

**FunctionPathInput** doesn need the "extension" inherited from **FunctionFileInput**, it'd be better to have:
**FunctionPathInput** => **FunctionFolderInput** 
**FunctionPathInput** => **FunctionFileInput**

Reply:
Sure. Will have a look and change this

Reply:
I read this again. FunctionPath actually need the extension in case the input is a file input. Path can be a file or a folder.